pandas/core/frame.py

```diff
@@ -3921,16 +3921,19 @@ def _set_value(
             Sets whether or not index/col interpreted as indexers
         """
         try:
+            # setitem will do validation that may raise TypeError,
+            #  ValueError, or LossySetitemError
             if takeable:
-                series = self._ixs(col, axis=1)
-                loc = index
+                icol = col
+                iindex = index
             else:
-                series = self._get_item_cache(col)
-                loc = self.index.get_loc(index)
-
-            # setitem_inplace will do validation that may raise TypeError,
-            #  ValueError, or LossySetitemError
-            series._mgr.setitem_inplace(loc, value)
+                icol = self.columns.get_loc(col)
+                iindex = self.index.get_loc(index)
+            # error: Argument 2 to "column_setitem" of "BlockManager" has
+            # incompatible type "Union[Hashable, Sequence[Hashable]]";
+            # expected "Union[int, slice, ndarray[Any, Any]]"
+            self._mgr.column_setitem(icol, iindex, value)  # type: ignore[arg-type]
+            self._clear_item_cache()
```

Comment:
could we make column_setitem always-inplace and make the clear_item_cache unecesssary?

API-wise i think the always-inplace method is a lot nicer than the less predictable one

Reply:
or maybe could make setitem_inplace ignore CoW since it is explicitly inplace?

Reply:
> could we make column_setitem always-inplace and make the clear_item_cache unecesssary?

The main use case of `column_setitem` is in `_iLocIndexer._setitem_single_column`, which is used for setting with `loc`/`iloc`. And for that use case, we need this to be not inplace (i.e. having the dtype coercing behaviour), since that is what we need for loc/iloc.

The case here is only for `at`/`iat` setting. 

I could add a separate inplace version or add an inplace keyword to `column_setitem` that could be used here. That would keep the current logic more intact, but since we fallback to loc/iloc anyway when the inplace setitem fails, I am not sure it would actually be very useful.

> or maybe could make setitem_inplace ignore CoW since it is explicitly inplace?

Even something that is explicitly inplace from a usage perspective will need to take care of CoW in https://github.com/pandas-dev/pandas/pull/46958

Reply:
@jbrockmendel do you have a preference on keeping this change with the `inplace` keyword for `column_setitem` or not?

Reply:
if you remove the TypeError/ValueError/LossySetitemError catching below, then ill have no objection to the non-inplace version

Reply:
Actually it seems we still need to catch TypeError, for example for a MultiIndex case where `self.columns.get_loc(col)` might not necessarily result in an integer.

So only removed LossySetitemError for now (and added a test case for setting with `at` with a MultiIndex, as that didn't yet seem to be covered in the indexing tests)

Reply:
OK, can you add a comment about why TypeError is needed.  what about ValueError?

Reply:
Will add a comment about TypeError. 
I am not fully sure about the ValueError. Is it possible for a setitem operation to raise a ValueError? (it seems that validation methods (like `_validate_setitem_value`) will mostly raise TypeErrors?) 

Now, this catching of a ValueError alraedy was here before, so I am hesitant to remove it without looking further in detail at it. I would prefer to leave that for another PR.

Reply:
So there are some cases where setitem actually raises a ValueError, eg when setting with an array-like that is not of length 1 (in this case of scalar indexer `at` or `iat`). 
Now, the fallback to loc/iloc will then most likely also raise a ValueError (so catching it might not necessarily add that much). But at least in some cases it seems to change the error message.

Reply:
For example:

```
>>> df = pd.DataFrame({'a': pd.Series([1, 2, 3], dtype="Int64"), 'b': [0.1, 0.2, 0.3]})
>>> df.iat[0, 0] = np.array([0, 1])
...
ValueError: Must have equal len keys and value when setting with an iterable
```

Without catching/reraising in iloc, the error would be "ValueError: setting an array element with a sequence", which is slightly less informative.

I suppose we should actually see to make this handling consistent in the different code paths (so it directly raises the more informative error message in the first place), but that's out of scope for this PR.